### PR TITLE
Support special identifier validation and repair

### PR DESCRIPTION
**Pull request by frankshen01**
# Description

Support for modifying special identifiers to meet Ansible's requirements, such as special identifiers beginning with a number.

# Type of change

- [x] New feature

# Testing

cat validation_identifiers.lst
```
{
    "hostname":"zhostname"
}
```

cat fortios_system_global_example.yml
```
- hosts: localhost
  vars:
   host: "192.168.52.177"
   username: "admin"
   password: "*********"
   vdom: "root"
   ssl_verify: "False"
  tasks:
  - name: Configure global attributes.
    fortios_system_global:
      host:  "{{ host }}"
      username: "{{ username }}"
      password: "{{ password }}"
      vdom:  "{{ vdom }}"
      https: "True"
      ssl_verify:  "{{ ssl_verify }}"
      system_global:
        admintimeout: "23"
        zhostname: "died111ie"

```

ansible-playbook fortios_system_global_example.yml
```
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] ****************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] **********************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Configure global attributes.] *********************************************************************************************************************************************************************************************************
changed: [localhost]

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

```
firewall config system global -> show
```
config system global
    set admin-server-cert "Server"
    set admintimeout 23
    set alias "FortiGate_VM64_KVM"
    set hostname "died111ie"
    set timezone 33
end

```

### CODE REVIEW

File: generate.py [hunk 1]
```diff
@@ -147,6 +179,8 @@ def jinjaExecutor(number=None):
     special_attributes_file = open('special_attributes.lst').read()
     special_attributes = json.loads(special_attributes_file)
 
+    validation_identifiers_file = open('validation_identifiers.lst').read()
+    validation_identifiers = json.loads(validation_identifiers_file)
```
Comment (mamunozgonzalez) [inline generate.py]: Can we call this 'valid_identifiers'? Would it make sense?

File: generate.py [hunk 2]
```diff
@@ -81,6 +110,9 @@ def renderModule(schema, version, special_attributes, version_added):
 
     schema['schema'] = hyphenToUnderscore(schema['schema'])
 
+    validation_identifiers_module = {}
```
Comment (mamunozgonzalez) [inline generate.py]: Why is there a difference between validation_identifiers_modules and validation_identifiers?

Comment (frankshen01) [reply generate.py]: Hi, @mamunozgonzalez Thanks for your review.  validation_identifiers includes all the attributes read from the validation_identifiers.lst file. validation_identifiers_modules only includes attributes to be modified in the current module. When the number of attributes from validation_identifiers.lst is large, there is no need to load all these attributes in the every module file. It only loads the attributes that exist in the current module. If the current module does not have any attributes to modify, the validation_identifiers_modules for the current module will be empty.

File: generate.py (hunk 1 above)
Comment (frankshen01) [reply generate.py]: Agree, fixed

File: ansible_templates/code.j2 [hunk 3]
```diff
@@ -62,7 +62,31 @@ def underscore_to_hyphen(data):
 
     return data
 
+{% if vi|length > 0 -%}
+def specilchar_to_originalchar(data):
```
[the remark this answers is not in the document]
Comment (frankshen01) [reply ansible_templates/code.j2]: Thank you,  fixed it.

File: validation_identifiers.lst [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "3gpp_plmn":"set_3gpp_plmn",
+    "0, 1 or 2":"set_0, 1 or 2",
```
[the remark this answers is not in the document]
Comment (frankshen01) [reply validation_identifiers.lst]: Hi @mamunozgonzalez , thanks, you are right,  after re-checking again,  I found the "0, 1 or 2" attribute is located in the diagnose__tree__.ipv6 file. It doesn't seem to be generated automatically, so "0, 1 or 2" attribute  will be removed from the validation_identifiers.lst.

File: ansible_templates/code.j2 [hunk 5]
```diff
@@ -62,7 +62,31 @@ def underscore_to_hyphen(data):
...
+            "{{k}}": "{{v}}"{%- if not loop.last %}, {%endif%}{% endfor -%}
+    }
+
+    for k,v in specillist.items():
+        if v == data:
+            return k
+
+    return data
+{% endif %}
+{%- if vi|length > 0 %}
```
[the remark this answers is not in the document]
Comment (frankshen01) [reply ansible_templates/code.j2]: Fixed it.

**Comment by frankshen01** [comment]
> Thanks for the PR @frankshen01.
> 
> Can you ensure that a full run does not modify the rest of the files? It would be as easy as generating everything and checking differences.

Hi, @mamunozgonzalez Thank you for your careful review. I made a comparison between the files generated before the change and the files generated after the change with Beyond Compare, Except for the file that needs to be modified, every other files remain the same. Below is the comparison report:

```
Folder Compare
Produced: 2019/9/9 22:34:47

Mode:  Differences
Left base folder: X:\ag\ansible_generator1
Right base folder: X:\ag\ansible_generator2
Name                                                            Size       Modified             Name                                                            Size       Modified
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
.git (h)                                                           410,711 2019/9/9 16:37:23    .git (h)                                                           417,635 2019/9/9 22:00:25
ansible_templates                                                   28,070 2019/9/9 11:29:34    ansible_templates                                                   29,158 2019/9/9 22:00:25
output                                                          12,160,724 2019/9/9 22:01:41    output                                                          12,162,926 2019/9/9 22:02:35
\v6.0.4                                                         12,160,724 2019/9/9 22:03:01    \v6.0.4                                                         12,162,926 2019/9/9 22:05:05
 +switch_controller                                                437,312 2019/9/9 22:02:58     +switch_controller                                                438,838 2019/9/9 22:05:02
 |+fortios_switch_controller_lldp_profile.py                        16,790 2019/9/9 22:19:32 <<  |+fortios_switch_controller_lldp_profile.py                        17,504 2019/9/9 22:21:20
 |+fortios_switch_controller_lldp_profile_example.yml                1,066 2019/9/9 22:02:52 <<  |+fortios_switch_controller_lldp_profile_example.yml                1,078 2019/9/9 22:04:55
 |+fortios_switch_controller_managed_switch.py                      64,036 2019/9/9 22:22:04 <<  |+fortios_switch_controller_managed_switch.py                      64,712 2019/9/9 22:23:53
 |+fortios_switch_controller_managed_switch_example.yml              5,781 2019/9/9 22:02:55 <<  |+fortios_switch_controller_managed_switch_example.yml              5,785 2019/9/9 22:04:58
 |\test_fortios_switch_controller_lldp_profile.py                   10,513 2019/9/9 22:20:06 <<  |\test_fortios_switch_controller_lldp_profile.py                   10,633 2019/9/9 22:21:53
 \wireless_controller_hotspot20                                    313,007 2019/9/9 22:32:13     \wireless_controller_hotspot20                                    313,683 2019/9/9 22:33:02
  +fortios_wireless_controller_hotspot20_hs_profile.py              26,056 2019/9/9 22:27:02 <<   +fortios_wireless_controller_hotspot20_hs_profile.py              26,688 2019/9/9 22:28:46
  +fortios_wireless_controller_hotspot20_hs_profile_example.yml      2,315 2019/9/9 22:02:02 <<   +fortios_wireless_controller_hotspot20_hs_profile_example.yml      2,319 2019/9/9 22:03:18
  \test_fortios_wireless_controller_hotspot20_hs_profile.py         20,777 2019/9/9 22:25:55 <<   \test_fortios_wireless_controller_hotspot20_hs_profile.py         20,817 2019/9/9 22:27:37
generate.py                                                          8,568 2019/9/9 22:01:19 >> generate.py                                                         10,197 2019/9/9 21:53:58
                                                                                             << valid_identifiers.lst                                                  182 2019/9/9 18:43:54
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


```

**Comment by mamunozgonzalez** [comment]
LGTM

**Pull request merged**